std/array.d

```diff
@@ -3891,3 +3891,231 @@ unittest
+nothrow:
+@safe:
+pure:
+
+/++
+Returns a static array constructed from `a`. The type of elements can be
+specified implicitly (`int[2] a = staticArray(1,2);`) or explicitly
+(`float[2] a = staticArray!float(1,2)`).
+D20180214T203015: The result is an rvalue, therefore uses like
+`foo(staticArray(1, 2, 3))` may be inefficient because of the copies.
++/
+// D20180214T185602: Workaround https://issues.dlang.org/show_bug.cgi?id=16779 (make alias to staticArray once fixed)
+pragma(inline, true) U[T.length] staticArray(U = CommonType!T, T...)(T a) @nogc
+{
+    return [a];
+}
+
+/// ditto
+pragma(inline, true) U[T.length] staticArrayCast(U, T...)(T a) @nogc
+{
+    enum n = T.length;
+    U[n] ret = void;
+    static foreach (i; 0 .. n)
+    {
+        // TODO: If any of these throws, the destructors of the already-constructed elements is not called. https://github.com/dlang/phobos/pull/4936/files#r131709329
+        emplaceRef!U(ret[i], cast(U) a[i]);
+    }
+    return ret;
+}
+
+///
+@safe unittest
+{
+    int val = 3;
+    staticArray(1, 2, val).assertSame!int([1, 2, 3]);
+    staticArray(1, 2.0).assertSame!double([1, 2.0]);
+    assert(!__traits(compiles, staticArray(1, "")));
+    staticArray().assertSame!void([]);
+    staticArray!float(1, 2).assertSame!float([1, 2]);
+    // see D20180214T185602
+    // auto a = staticArray!byte(1, 2);
+    staticArrayCast!byte(1, 2).assertSame!byte([1, 2]);
+    staticArrayCast!byte(1, 129).assertSame!byte([1, -127]);
+
+    staticArrayCast!(const(int))(1, 2).assertSame!(const(int))([1, 2]);
+    staticArrayCast!(immutable(int))(1, 2).assertSame!(immutable(int))([1, 2]);
+    staticArray([1]).assertSame!(int[])([[1]]);
+    version (Bug)
+    {
+        // NOTE: correctly issues a deprecation
+        int[] a2 = staticArray(1, 2);
+    }
+}
+
+/++
+Returns a static array constructed from `arr`. The type of elements can be
+specified implicitly (`int[2] a = [1,2].asStatic;`) or explicitly
+(`float[2] a = [1,2].asStaticCast!float`).
+See also D20180214T203015.
++/
+pragma(inline, true) T[n] asStatic(T, size_t n)(auto ref T[n] arr) @nogc
+{
+    return arr;
+}
+
+/// ditto
+U[n] asStaticCast(U, T, size_t n)(auto ref T[n] arr) @nogc
+{
+    U[n] ret = void;
+    static foreach (i; 0 .. n)
+    {
+        emplaceRef!U(ret[i], cast(U) arr[i]);
+    }
+    return ret;
+}
+
+///
+@safe unittest
+{
+    int val = 3;
+    static immutable gold = [1, 2, 3];
+    [1, 2, val].asStatic.assertSame!int([1, 2, 3]);
+
+    @nogc void checkNogc()
+    {
+        [1, 2, val].asStatic.assertSame!int(gold);
+    }
+
+    [1, 2, val].asStaticCast!double.assertSame!double(gold);
+    [1, 2, 3].asStaticCast!int.assertSame!int(gold);
+
+    [1, 2, 3].asStaticCast!(const(int)).assertSame!(const(int))(gold);
+    [1, 2, 3].asStaticCast!(const(double)).assertSame!(const(double))(gold);
+    {
+        const(int)[3] a2 = [1, 2, 3].asStatic;
+    }
+
+    [1, 129].asStaticCast!byte.assertSame!byte([1, -127]);
+
+    version (Bug)
+    {
```

Comment:
This shouldn't be part of the __public__ documentation

Reply:
done